sw/device/tests/rv_core_ibex_nmi_irq_test.c

```diff
@@ -0,0 +1,271 @@
+
+  // Enables alert handler irq.
+  CHECK_DIF_OK(dif_alert_handler_irq_set_enabled(
+      &alert_handler, kDifAlertHandlerIrqClassa, kDifToggleEnabled));
+}
+
+/**
+ * Handle the wdog bark nmi irq.
+ */
+static void wdog_handler(void) {
+  top_earlgrey_plic_peripheral_t peripheral;
+  dif_aon_timer_irq_t irq;
+  isr_testutils_aon_timer_isr(plic_ctx, aon_timer_ctx, &peripheral, &irq);
+  CHECK_DIF_OK(dif_aon_timer_watchdog_stop(&aon_timer));
+}
+
+/**
+ * Handle the alert handler nmi irq.
+ */
+static void alert_handler_handler(void) {
+  top_earlgrey_plic_peripheral_t peripheral;
+  dif_alert_handler_irq_t irq;
+  isr_testutils_alert_handler_isr(plic_ctx, alert_handler_ctx, &peripheral,
+                                  &irq);
+
+  // Check that the alert handler escalation can be stoped.
+  bool can_clear;
+  CHECK_DIF_OK(dif_alert_handler_escalation_can_clear(
+      alert_handler_ctx.alert_handler, kDifAlertHandlerClassA, &can_clear));
+  CHECK(can_clear, "Alert handler is locked.");
+
+  CHECK_DIF_OK(dif_alert_handler_escalation_clear(
+      alert_handler_ctx.alert_handler, kDifAlertHandlerClassA));
+}
+
+/**
+ * Execute the wdog bark nmi interrupt test.
+ */
+static void wdog_nmi_test(void) {
+  rv_plic_testutils_irq_range_enable(
+      &plic, plic_ctx.hart_id, kTopEarlgreyPlicIrqIdAonTimerAonWdogTimerBark,
+      kTopEarlgreyPlicIrqIdAonTimerAonWdogTimerBark);
+
+  CHECK_DIF_OK(
+      dif_rv_core_ibex_enable_nmi(&rv_core_ibex, kDifRvCoreIbexNmiSourceWdog));
+
+  // Setup to handle the incoming wdog nmi irq.
+  expected_isr_handler = wdog_handler;
+  nmi_fired = false;
+  nmi_state = (dif_rv_core_ibex_nmi_state_t){0};
+
+  // Setup the wdog bark interrupt.
+  uint32_t count_cycles =
+      aon_timer_testutils_get_aon_cycles_from_us(kWdogBarkMicros);
+  aon_timer_testutils_watchdog_config(&aon_timer,
+                                      /*bark_cycles=*/count_cycles,
+                                      /*bite_cycles=*/UINT32_MAX,
+                                      /*pause_in_sleep=*/false);
+  IBEX_SPIN_FOR(nmi_fired, kWdogBarkMicros * 2);
+
+  // Check that the wdog nmi was enabled and barked during the nmi irq.
+  CHECK(nmi_state.wdog_enabled && nmi_state.wdog_barked,
+        "Wdog nmi state not expected:\n\t"
+        "wdog_enable:%x\n\twdog_raised:%"
+        "x\n",
+        nmi_state.wdog_enabled, nmi_state.wdog_barked);
+
+  // Check that the wdog nmi is enable and is cleared.
+  CHECK_DIF_OK(dif_rv_core_ibex_get_nmi_state(
+      &rv_core_ibex, (dif_rv_core_ibex_nmi_state_t *)&nmi_state));
+  CHECK(nmi_state.wdog_enabled && !nmi_state.wdog_barked,
+        "Wdog nmi state not expected:\n\t"
+        "wdog_enable:%x\n\twdog_raised:%"
+        "x\n",
+        nmi_state.wdog_enabled, nmi_state.wdog_barked);
+}
+
+/**
+ * Execute the alert handler nmi interrupt test.
+ */
+static void alert_handler_nmi_test(void) {
+  alert_handler_config();
+  rv_plic_testutils_irq_range_enable(&plic, plic_ctx.hart_id,
```

Comment:
do we need to enable these..? we're not trying to test the plic interrupts in this test right? 
if we are, i'm a bit surprised how the test is passing right now because i don't think i see a normal irq handler.

Reply:
Yes, without it the nmi didn't fire. I'm assuming that the nmi is taking place of the external irq.

Reply:
eh?! what the ... that's not right (i don't think).
This is probably something we need to investigate. The nmi path and the plic path are technically supposed to be different, unless there's something in `ottf` that links them.